### ISSUE TITLE
Reduce default concurrency

### ISSUE THREAD

**Issue by mperham** [mperham/sidekiq#3892]
Today Sidekiq uses a default concurrency of 25.  These means Sidekiq will spawn 25 worker threads and execute up to 25 jobs concurrently in a process.

glibc has a major memory fragmentation issue which gets worse with more threads, causing many people to move to jemalloc.

I also happen to think that, with time and experience but no hard data, that 25 is pretty aggressive and most apps can peg a CPU with a lower number of cores.  Developers testing locally on macOS rarely need such large concurrency.

I'd suggest we reduce the default concurrency from 25 to 15 in Sidekiq 5.2.0.  This will save memory and reduce fragmentation and bloat on Linux.  Anyone who wants to retain the old value can add `-c 25` to their command line.

WDYT?

**Comment by jc00ke** [comment]
I'd be curious to see the hard data. I don't think it would hurt to lower it, and my guess is that anyone that needs more than 15 will knowingly tune it to a much higher number in an environment where they have more processing power (ie not most Heroku dynos.)

**Comment by amcaplan** [comment]
Don't think I've ever seen a need for >10. 4-6 workers per process, with 1 process per core, is far more common in my experience.

**Comment by schneems** [comment]
Heroku has plenty of processing power thank you very much. The “performance” and “private” large dyno has 14GB of ram and 8 dedicated VCPUs (8 hyperthreads backed by 4 real cores on top of a hypervisor). Though I do realize you said “most”.

FWIW I think 25 is pretty high. Puma default is 16. Even then most people tune it down on the web. It would be helpful to get some kid of a standardized metric around when it is helpful to add extra sidekiq workers on a box.

**Comment by noahgibbs** [comment]
I can't tell here if you mean per process, or total. For total across multiple processes, 25 is probably great. Per process, yeah, 5 is reasonable, 10 is high and 25 is *very* high. Given the GIL, it's very hard to get CRuby to productively use more than 10-ish threads for real tasks -- and in cases where you can, it's because something like EventMachine or Node.js would have been a better choice than Ruby threads.

**Comment by mperham** [comment]
I could be talked down to 10.  I think 5 is too low; most business apps are I/O heavy, allowing pretty decent concurrency, even with GIL.

**Comment by jagthedrummer** [comment]
10 sounds about right to me, and even 15 would be better than 25.

**Comment by zachmccormick** [comment]
Is there a nice way to use `Etc.nprocessors` to determine a happy number for most people? I think Celery uses that sort of logic to determine default concurrency FWIW. I do agree with those above that in optimized cases you're going to want to look at the nature of your workload and tweak accordingly.

**Comment by amcaplan** [comment]
@zachmccormick number of processors is irrelevant for MRI (i.e. what the vast majority of the community uses AFAICT); each Sidekiq process can be handled by only 1 processor due to the GIL, since Sidekiq achieves concurrency via threads.  You need to run multiple Sidekiq processes to achieve true parallel processing.

**Comment by zachmccormick** [comment]
Ah I see - didn't realize that! Thanks!

**Comment by mperham** [comment]
Sidekiq doesn't fork or scale processes, only threads.  You need to start multiple Sidekiqs yourself, using the tool/init of your choice.  Sidekiq Enterprise has a multi-process `sidekiqswarm` binary which scales Sidekiq processes according to CPU count.

https://github.com/mperham/sidekiq/wiki/Ent-Multi-Process

**Comment by noahgibbs** [comment]
@amcaplan @zachmccormick The number of processors matters if you add processes. But as @mperham says, you'd have to do that yourself - Sidekiq won't do that automatically.

Also, while the GIL means each Ruby thread blocks all others *when running Ruby*, there are some non-Ruby operations (e.g. network or disk I/O, database, some parts of garbage collection, many things done with native extensions) which can happen on a background thread and don't block your Ruby process. Those things can happen in parallel if you have more than one processor, but not if you don't. That's a lot of what I was talking about above with "saturating" a processor with 6+ threads per process - that makes sure that even when most of your Ruby code is blocked, *something* is running and making forward progress.

**Comment by noahgibbs** [comment]
Yup! That's a great summary. But since the CPU percentage for a given task can vary, there's a bit of an asymptote as far as how many threads are necessary to saturate...

**Comment by mperham** [comment]
It's obvious that glibc's memory bloat, [as discussed on my blog](https://www.mikeperham.com/2018/04/25/taming-rails-memory-bloat/), gets worse as concurrency increases.  I think reducing concurrency from 25 to 10 will reduce memory usage AND bloat, giving us a double win in memory.

Pro tip: you can get the old behavior by adding `-c 25` to the command line.
